### PR TITLE
Refs #6611: Bugfix/throughput wait and sync [6678]

### PR DESCRIPTION
**Pull request by EduPonz**
Publisher
    * Discover data subscriber before sending TEST_STARTS command
    * Wait for ack when sending TEST_STARTS
    * Only wait whatever is left of `recovery_time`
Subscriber
    * wait_for_unread_samples instead of attempting to `take` all the time

### CODE REVIEW

File: test/performance/ThroughputPublisher.cpp [hunk 1]
```diff
@@ -509,27 +509,42 @@ bool ThroughputPublisher::test(
...
 
-    t_end_ = std::chrono::steady_clock::now();
+    std::unique_lock<std::mutex> data_disc_lock(dataMutex_);
+    data_disc_cond_.wait(data_disc_lock, [&]()
+    {
+        return data_disc_count_ > 0;
+    });
+    data_disc_lock.unlock();
 
+    // Declate test time variables
```
Comment (MiguelCompany) [inline test/performance/ThroughputPublisher.cpp]:
```suggestion
    // Declare test time variables
```

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/724/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/617/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/719/)
* Performance [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Performance%20Tests%20Localhost/189/)

File: test/performance/ThroughputPublisher.cpp [hunk 2]
```diff
@@ -543,10 +558,19 @@ bool ThroughputPublisher::test(
...
         samples += demand;
-        std::this_thread::sleep_for(std::chrono::milliseconds(recovery_time_ms));
-        timewait_us += t_overhead_;
+
+        // If the batch took less than the recovery time, sleep for the difference recovery_time - batch_duration.
+        // Else, go ahead with next batch without time to recover.
+        if ((t_end_ - batch_start) <= std::chrono::milliseconds(recovery_time_ms))
+        {
+            std::this_thread::sleep_for(std::chrono::milliseconds(recovery_time_ms) - (t_end_ - batch_start));
+        }
```
Comment (MiguelCompany) [inline test/performance/ThroughputPublisher.cpp]: I would calculate the time to sleep first, and then do the sleep only if that time is greater than 0.

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/729/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/622/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/725/)
* Performance [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Performance%20Tests%20Localhost/194/)

**Comment by MiguelCompany** [comment]
Failed tests not related. Merging.

**Pull request merged**